terraform test expect_failures does not capture output precondition failure on apply

### Terraform Version

```shell
Terraform v1.6.0-beta1
on windows_amd64
```


### Terraform Configuration Files

main.tf
```terraform
variable "foo" {
  type    = bool
  default = true
}

output "bar" {
  value = var.foo

  precondition {
    condition     = false && var.foo
    error_message = "Always fails"
  }
}
```

apply.tftest.hcl
```hcl
run "apply" {
  command = apply

  expect_failures = [
    output.bar
  ]
}
```

plan.tftest.hcl
```hcl
run "plan" {
  command = plan

  expect_failures = [
    output.bar
  ]
}
```

### Debug Output
```
2023-09-09T23:10:36.674+0100 [INFO]  Terraform version: 1.6.0 beta1
2023-09-09T23:10:36.680+0100 [DEBUG] using github.com/hashicorp/go-tfe v1.32.1
2023-09-09T23:10:36.680+0100 [DEBUG] using github.com/hashicorp/hcl/v2 v2.18.0
2023-09-09T23:10:36.680+0100 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2023-09-09T23:10:36.680+0100 [DEBUG] using github.com/zclconf/go-cty v1.14.0
2023-09-09T23:10:36.680+0100 [INFO]  Go runtime version: go1.21.0
2023-09-09T23:10:36.680+0100 [INFO]  CLI args: []string{"C:\\Users\\sweeand\\dev\\terraform\\1.6.0-beta1\\terraform.exe", "test"}
2023-09-09T23:10:36.680+0100 [TRACE] Stdout is not a terminal
2023-09-09T23:10:36.680+0100 [TRACE] Stderr is not a terminal
2023-09-09T23:10:36.680+0100 [TRACE] Stdin is a terminal
2023-09-09T23:10:36.688+0100 [DEBUG] Attempting to open CLI config file: C:\Users\sweeand\AppData\Roaming\terraform.rc
2023-09-09T23:10:36.692+0100 [INFO]  Loading CLI configuration from C:\Users\sweeand\AppData\Roaming\terraform.rc
2023-09-09T23:10:36.696+0100 [INFO]  Loading CLI configuration from C:\Users\sweeand\AppData\Roaming\terraform.d\credentials.tfrc.json
2023-09-09T23:10:36.711+0100 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2023-09-09T23:10:36.711+0100 [DEBUG] ignoring non-existing provider search directory C:\Users\sweeand\AppData\Roaming\terraform.d\plugins
2023-09-09T23:10:36.715+0100 [DEBUG] ignoring non-existing provider search directory C:\Users\sweeand\AppData\Roaming\HashiCorp\Terraform\plugins
2023-09-09T23:10:36.715+0100 [INFO]  CLI command args: []string{"test"}
2023-09-09T23:10:36.757+0100 [DEBUG] checking for provisioner in "."
2023-09-09T23:10:36.761+0100 [DEBUG] checking for provisioner in "C:\\Users\\sweeand\\dev\\terraform\\1.6.0-beta1"
2023-09-09T23:10:36.770+0100 [DEBUG] TestSuiteRunner: found 2 files with 2 run blocks
2023-09-09T23:10:36.770+0100 [TRACE] TestFileRunner: executing test file apply.tftest.hcl
2023-09-09T23:10:36.770+0100 [TRACE] TestFileRunner: executing run block apply.tftest.hcl/apply
2023-09-09T23:10:36.770+0100 [TRACE] TestFileRunner: called validate for apply.tftest.hcl/apply
2023-09-09T23:10:36.770+0100 [TRACE] terraform.NewContext: starting
2023-09-09T23:10:36.770+0100 [TRACE] terraform.NewContext: complete
2023-09-09T23:10:36.770+0100 [TRACE] TestFileRunner: waiting for execution during apply.tftest.hcl/apply
2023-09-09T23:10:36.770+0100 [DEBUG] TestFileRunner: starting validate for apply.tftest.hcl/apply
2023-09-09T23:10:36.770+0100 [DEBUG] Building and walking validate graph
2023-09-09T23:10:36.770+0100 [TRACE] building graph for walkValidate
2023-09-09T23:10:36.770+0100 [TRACE] Executing graph transform *terraform.ConfigTransformer
2023-09-09T23:10:36.770+0100 [TRACE] ConfigTransformer: Starting for path:
2023-09-09T23:10:36.770+0100 [TRACE] Completed graph transform *terraform.ConfigTransformer (no changes)
2023-09-09T23:10:36.770+0100 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2023-09-09T23:10:36.770+0100 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.770+0100 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2023-09-09T23:10:36.770+0100 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2023-09-09T23:10:36.770+0100 [TRACE] Executing graph transform *terraform.LocalTransformer
2023-09-09T23:10:36.770+0100 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2023-09-09T23:10:36.770+0100 [TRACE] Executing graph transform *terraform.OutputTransformer
2023-09-09T23:10:36.770+0100 [TRACE] OutputTransformer: adding bar as *terraform.nodeExpandOutput
2023-09-09T23:10:36.770+0100 [TRACE] Completed graph transform *terraform.OutputTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.770+0100 [TRACE] Executing graph transform *terraform.checkTransformer
2023-09-09T23:10:36.770+0100 [TRACE] Completed graph transform *terraform.checkTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.StateTransformer
2023-09-09T23:10:36.771+0100 [TRACE] StateTransformer: pointless no-op call, creating no nodes at all
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2023-09-09T23:10:36.771+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.graphTransformerMulti (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.ModuleExpansionTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.ModuleExpansionTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.ExternalReferenceTransformer
2023-09-09T23:10:36.771+0100 [TRACE] Completed graph transform *terraform.ExternalReferenceTransformer (no changes)
2023-09-09T23:10:36.771+0100 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2023-09-09T23:10:36.771+0100 [DEBUG] ReferenceTransformer: "var.foo" references: []
2023-09-09T23:10:36.772+0100 [DEBUG] ReferenceTransformer: "output.bar (expand)" references: [var.foo var.foo]
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.ReferenceTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.AttachDependenciesTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.AttachDependenciesTransformer (no changes)
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.attachDataResourceDependsOnTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.attachDataResourceDependsOnTransformer (no changes)
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.DestroyEdgeTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.DestroyEdgeTransformer (no changes)
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.pruneUnusedNodesTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.pruneUnusedNodesTransformer (no changes)
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.TargetsTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.ForcedCBDTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.ForcedCBDTransformer (no changes)
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.CloseProviderTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.CloseProviderTransformer (no changes)
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.CloseRootModuleTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.CloseRootModuleTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  root - *terraform.nodeCloseModule
    output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.772+0100 [TRACE] Executing graph transform *terraform.TransitiveReductionTransformer
2023-09-09T23:10:36.772+0100 [TRACE] Completed graph transform *terraform.TransitiveReductionTransformer (no changes)
2023-09-09T23:10:36.772+0100 [DEBUG] Starting graph walk: walkValidate
2023-09-09T23:10:36.772+0100 [TRACE] vertex "var.foo": starting visit (*terraform.NodeRootVariable)
2023-09-09T23:10:36.772+0100 [TRACE] NodeRootVariable: evaluating var.foo
2023-09-09T23:10:36.772+0100 [TRACE] prepareFinalInputVariableValue: preparing var.foo
2023-09-09T23:10:36.772+0100 [TRACE] prepareFinalInputVariableValue: var.foo has a default value
2023-09-09T23:10:36.772+0100 [TRACE] BuiltinEvalContext: Storing final value for variable var.foo
2023-09-09T23:10:36.773+0100 [TRACE] evalVariableValidations: no validation rules declared for var.foo, so skipping
2023-09-09T23:10:36.773+0100 [TRACE] vertex "var.foo": visit complete
2023-09-09T23:10:36.773+0100 [TRACE] vertex "output.bar (expand)": starting visit (*terraform.nodeExpandOutput)
2023-09-09T23:10:36.773+0100 [TRACE] vertex "output.bar (expand)": expanding dynamic subgraph
2023-09-09T23:10:36.773+0100 [TRACE] Expanding output: adding output.bar as *terraform.NodeApplyableOutput
2023-09-09T23:10:36.773+0100 [TRACE] vertex "output.bar (expand)": entering dynamic subgraph
2023-09-09T23:10:36.773+0100 [TRACE] vertex "output.bar": starting visit (*terraform.NodeApplyableOutput)
2023-09-09T23:10:36.773+0100 [TRACE] evalCheckRules: output.bar status is now StatusUnknown
2023-09-09T23:10:36.773+0100 [TRACE] setValue: Saving Create change for output.bar in changeset
2023-09-09T23:10:36.773+0100 [TRACE] setValue: Saving value for output.bar in state
2023-09-09T23:10:36.773+0100 [TRACE] vertex "output.bar": visit complete
2023-09-09T23:10:36.773+0100 [TRACE] vertex "root": starting visit (terraform.graphNodeRoot)
2023-09-09T23:10:36.773+0100 [TRACE] vertex "root": visit complete
2023-09-09T23:10:36.773+0100 [TRACE] vertex "output.bar (expand)": dynamic subgraph completed successfully
2023-09-09T23:10:36.774+0100 [TRACE] vertex "output.bar (expand)": visit complete
2023-09-09T23:10:36.774+0100 [TRACE] vertex "root": starting visit (*terraform.nodeCloseModule)
2023-09-09T23:10:36.774+0100 [TRACE] vertex "root": visit complete
2023-09-09T23:10:36.774+0100 [DEBUG] TestFileRunner: completed validate for  apply.tftest.hcl/apply
2023-09-09T23:10:36.774+0100 [TRACE] TestFileRunner: called plan for apply.tftest.hcl/apply
2023-09-09T23:10:36.774+0100 [TRACE] terraform.NewContext: starting
2023-09-09T23:10:36.774+0100 [TRACE] terraform.NewContext: complete
2023-09-09T23:10:36.774+0100 [TRACE] TestFileRunner: waiting for execution during apply.tftest.hcl/apply
2023-09-09T23:10:36.774+0100 [DEBUG] TestFileRunner: starting plan for apply.tftest.hcl/apply
2023-09-09T23:10:36.774+0100 [DEBUG] Building and walking plan graph for NormalMode
2023-09-09T23:10:36.774+0100 [TRACE] building graph for walkPlan
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.ConfigTransformer
2023-09-09T23:10:36.774+0100 [TRACE] ConfigTransformer: Starting for path:
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.ConfigTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.LocalTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.OutputTransformer
2023-09-09T23:10:36.774+0100 [TRACE] OutputTransformer: adding bar as *terraform.nodeExpandOutput
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.OutputTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.checkTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.checkTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.StateTransformer
2023-09-09T23:10:36.774+0100 [TRACE] StateTransformer: creating nodes for deposed instance objects only
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2023-09-09T23:10:36.774+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.graphTransformerMulti (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.ModuleExpansionTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.ModuleExpansionTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.ExternalReferenceTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.ExternalReferenceTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2023-09-09T23:10:36.774+0100 [DEBUG] ReferenceTransformer: "var.foo" references: []
2023-09-09T23:10:36.774+0100 [DEBUG] ReferenceTransformer: "output.bar (expand)" references: [var.foo var.foo]
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.ReferenceTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.AttachDependenciesTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.AttachDependenciesTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.attachDataResourceDependsOnTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.attachDataResourceDependsOnTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.DestroyEdgeTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.DestroyEdgeTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.pruneUnusedNodesTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.pruneUnusedNodesTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.TargetsTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.ForcedCBDTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.ForcedCBDTransformer (no changes)
2023-09-09T23:10:36.774+0100 [TRACE] Executing graph transform *terraform.CloseProviderTransformer
2023-09-09T23:10:36.774+0100 [TRACE] Completed graph transform *terraform.CloseProviderTransformer (no changes)
2023-09-09T23:10:36.775+0100 [TRACE] Executing graph transform *terraform.CloseRootModuleTransformer
2023-09-09T23:10:36.775+0100 [TRACE] Completed graph transform *terraform.CloseRootModuleTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  root - *terraform.nodeCloseModule
    output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.775+0100 [TRACE] Executing graph transform *terraform.TransitiveReductionTransformer
2023-09-09T23:10:36.775+0100 [TRACE] Completed graph transform *terraform.TransitiveReductionTransformer (no changes)
2023-09-09T23:10:36.775+0100 [DEBUG] Starting graph walk: walkPlan
2023-09-09T23:10:36.775+0100 [TRACE] vertex "var.foo": starting visit (*terraform.NodeRootVariable)
2023-09-09T23:10:36.775+0100 [TRACE] NodeRootVariable: evaluating var.foo
2023-09-09T23:10:36.775+0100 [TRACE] prepareFinalInputVariableValue: preparing var.foo
2023-09-09T23:10:36.775+0100 [TRACE] prepareFinalInputVariableValue: var.foo has a default value
2023-09-09T23:10:36.775+0100 [TRACE] prepareFinalInputVariableValue: var.foo has no defined value
2023-09-09T23:10:36.775+0100 [TRACE] BuiltinEvalContext: Storing final value for variable var.foo
2023-09-09T23:10:36.775+0100 [TRACE] evalVariableValidations: no validation rules declared for var.foo, so skipping
2023-09-09T23:10:36.775+0100 [TRACE] vertex "var.foo": visit complete
2023-09-09T23:10:36.775+0100 [TRACE] vertex "output.bar (expand)": starting visit (*terraform.nodeExpandOutput)
2023-09-09T23:10:36.775+0100 [TRACE] vertex "output.bar (expand)": expanding dynamic subgraph
2023-09-09T23:10:36.775+0100 [TRACE] Expanding output: adding output.bar as *terraform.NodeApplyableOutput
2023-09-09T23:10:36.775+0100 [TRACE] vertex "output.bar (expand)": entering dynamic subgraph
2023-09-09T23:10:36.775+0100 [TRACE] vertex "output.bar": starting visit (*terraform.NodeApplyableOutput)
2023-09-09T23:10:36.775+0100 [TRACE] evalCheckRules: output.bar status is now StatusFail
2023-09-09T23:10:36.775+0100 [ERROR] vertex "output.bar" error: Module output value precondition failed
2023-09-09T23:10:36.775+0100 [TRACE] vertex "output.bar": visit complete, with errors
2023-09-09T23:10:36.775+0100 [TRACE] dag/walk: upstream of "root" errored, so skipping
2023-09-09T23:10:36.775+0100 [TRACE] vertex "output.bar (expand)": dynamic subgraph encountered errors: Module output value precondition failed
2023-09-09T23:10:36.775+0100 [ERROR] vertex "output.bar (expand)" error: Module output value precondition failed
2023-09-09T23:10:36.775+0100 [TRACE] vertex "output.bar (expand)": visit complete, with errors
2023-09-09T23:10:36.775+0100 [TRACE] dag/walk: upstream of "root" errored, so skipping
2023-09-09T23:10:36.775+0100 [DEBUG] TestFileRunner: completed plan for apply.tftest.hcl/apply
apply.tftest.hcl... [31mfail[0m[0m
  run "apply"... [31mfail[0m[0m
[31m╷[0m[0m
[31m│[0m [0m[1m[31mError: [0m[0m[1mModule output value precondition failed[0m
[31m│[0m [0m
[31m│[0m [0m[0m  on main.tf line 10, in output "bar":
[31m│[0m [0m  10:     condition     = [4mfalse && var.foo[0m[0m
[31m│[0m [0m    [90m├────────────────[0m
[31m│[0m [0m[0m    [90m│[0m [1mvar.foo[0m is true
[31m│[0m [0m[0m
[31m│[0m [0mAlways fails
[31m╵[0m[0m
2023-09-09T23:10:36.776+0100 [TRACE] TestStateManager: cleaning up state for apply.tftest.hcl
2023-09-09T23:10:36.776+0100 [TRACE] TestFileRunner: executing test file plan.tftest.hcl
2023-09-09T23:10:36.776+0100 [TRACE] TestFileRunner: executing run block plan.tftest.hcl/plan
2023-09-09T23:10:36.776+0100 [TRACE] TestFileRunner: called validate for plan.tftest.hcl/plan
2023-09-09T23:10:36.776+0100 [TRACE] terraform.NewContext: starting
2023-09-09T23:10:36.776+0100 [TRACE] terraform.NewContext: complete
2023-09-09T23:10:36.776+0100 [TRACE] TestFileRunner: waiting for execution during plan.tftest.hcl/plan
2023-09-09T23:10:36.776+0100 [DEBUG] TestFileRunner: starting validate for plan.tftest.hcl/plan
2023-09-09T23:10:36.776+0100 [DEBUG] Building and walking validate graph
2023-09-09T23:10:36.776+0100 [TRACE] building graph for walkValidate
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.ConfigTransformer
2023-09-09T23:10:36.776+0100 [TRACE] ConfigTransformer: Starting for path:
2023-09-09T23:10:36.776+0100 [TRACE] Completed graph transform *terraform.ConfigTransformer (no changes)
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2023-09-09T23:10:36.776+0100 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2023-09-09T23:10:36.776+0100 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.LocalTransformer
2023-09-09T23:10:36.776+0100 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.OutputTransformer
2023-09-09T23:10:36.776+0100 [TRACE] OutputTransformer: adding bar as *terraform.nodeExpandOutput
2023-09-09T23:10:36.776+0100 [TRACE] Completed graph transform *terraform.OutputTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.checkTransformer
2023-09-09T23:10:36.776+0100 [TRACE] Completed graph transform *terraform.checkTransformer (no changes)
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2023-09-09T23:10:36.776+0100 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2023-09-09T23:10:36.776+0100 [TRACE] Executing graph transform *terraform.StateTransformer
2023-09-09T23:10:36.776+0100 [TRACE] StateTransformer: pointless no-op call, creating no nodes at all
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2023-09-09T23:10:36.777+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.graphTransformerMulti (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.ModuleExpansionTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.ModuleExpansionTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.ExternalReferenceTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.ExternalReferenceTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2023-09-09T23:10:36.777+0100 [DEBUG] ReferenceTransformer: "var.foo" references: []
2023-09-09T23:10:36.777+0100 [DEBUG] ReferenceTransformer: "output.bar (expand)" references: [var.foo var.foo]
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.ReferenceTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.AttachDependenciesTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.AttachDependenciesTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.attachDataResourceDependsOnTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.attachDataResourceDependsOnTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.DestroyEdgeTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.DestroyEdgeTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.pruneUnusedNodesTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.pruneUnusedNodesTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.TargetsTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.ForcedCBDTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.ForcedCBDTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.CloseProviderTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.CloseProviderTransformer (no changes)
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.CloseRootModuleTransformer
2023-09-09T23:10:36.777+0100 [TRACE] Completed graph transform *terraform.CloseRootModuleTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  root - *terraform.nodeCloseModule
    output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.777+0100 [TRACE] Executing graph transform *terraform.TransitiveReductionTransformer
2023-09-09T23:10:36.778+0100 [TRACE] Completed graph transform *terraform.TransitiveReductionTransformer (no changes)
2023-09-09T23:10:36.778+0100 [DEBUG] Starting graph walk: walkValidate
2023-09-09T23:10:36.778+0100 [TRACE] vertex "var.foo": starting visit (*terraform.NodeRootVariable)
2023-09-09T23:10:36.778+0100 [TRACE] NodeRootVariable: evaluating var.foo
2023-09-09T23:10:36.778+0100 [TRACE] prepareFinalInputVariableValue: preparing var.foo
2023-09-09T23:10:36.778+0100 [TRACE] prepareFinalInputVariableValue: var.foo has a default value
2023-09-09T23:10:36.778+0100 [TRACE] BuiltinEvalContext: Storing final value for variable var.foo
2023-09-09T23:10:36.778+0100 [TRACE] evalVariableValidations: no validation rules declared for var.foo, so skipping
2023-09-09T23:10:36.778+0100 [TRACE] vertex "var.foo": visit complete
2023-09-09T23:10:36.778+0100 [TRACE] vertex "output.bar (expand)": starting visit (*terraform.nodeExpandOutput)
2023-09-09T23:10:36.778+0100 [TRACE] vertex "output.bar (expand)": expanding dynamic subgraph
2023-09-09T23:10:36.778+0100 [TRACE] Expanding output: adding output.bar as *terraform.NodeApplyableOutput
2023-09-09T23:10:36.778+0100 [TRACE] vertex "output.bar (expand)": entering dynamic subgraph
2023-09-09T23:10:36.778+0100 [TRACE] vertex "output.bar": starting visit (*terraform.NodeApplyableOutput)
2023-09-09T23:10:36.778+0100 [TRACE] evalCheckRules: output.bar status is now StatusUnknown
2023-09-09T23:10:36.778+0100 [TRACE] setValue: Saving Create change for output.bar in changeset
2023-09-09T23:10:36.778+0100 [TRACE] setValue: Saving value for output.bar in state
2023-09-09T23:10:36.778+0100 [TRACE] vertex "output.bar": visit complete
2023-09-09T23:10:36.778+0100 [TRACE] vertex "root": starting visit (terraform.graphNodeRoot)
2023-09-09T23:10:36.778+0100 [TRACE] vertex "root": visit complete
2023-09-09T23:10:36.778+0100 [TRACE] vertex "output.bar (expand)": dynamic subgraph completed successfully
2023-09-09T23:10:36.778+0100 [TRACE] vertex "output.bar (expand)": visit complete
2023-09-09T23:10:36.779+0100 [TRACE] vertex "root": starting visit (*terraform.nodeCloseModule)
2023-09-09T23:10:36.779+0100 [TRACE] vertex "root": visit complete
2023-09-09T23:10:36.779+0100 [DEBUG] TestFileRunner: completed validate for  plan.tftest.hcl/plan
2023-09-09T23:10:36.779+0100 [TRACE] TestFileRunner: called plan for plan.tftest.hcl/plan
2023-09-09T23:10:36.779+0100 [TRACE] terraform.NewContext: starting
2023-09-09T23:10:36.779+0100 [TRACE] terraform.NewContext: complete
2023-09-09T23:10:36.779+0100 [TRACE] TestFileRunner: waiting for execution during plan.tftest.hcl/plan
2023-09-09T23:10:36.779+0100 [DEBUG] TestFileRunner: starting plan for plan.tftest.hcl/plan
2023-09-09T23:10:36.779+0100 [DEBUG] Building and walking plan graph for NormalMode
2023-09-09T23:10:36.779+0100 [TRACE] building graph for walkPlan
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.ConfigTransformer
2023-09-09T23:10:36.779+0100 [TRACE] ConfigTransformer: Starting for path:
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.ConfigTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.LocalTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.OutputTransformer
2023-09-09T23:10:36.779+0100 [TRACE] OutputTransformer: adding bar as *terraform.nodeExpandOutput
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.OutputTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.checkTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.checkTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.StateTransformer
2023-09-09T23:10:36.779+0100 [TRACE] StateTransformer: creating nodes for deposed instance objects only
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2023-09-09T23:10:36.779+0100 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.graphTransformerMulti (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.ModuleExpansionTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.ModuleExpansionTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.ExternalReferenceTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.ExternalReferenceTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2023-09-09T23:10:36.779+0100 [DEBUG] ReferenceTransformer: "var.foo" references: []
2023-09-09T23:10:36.779+0100 [DEBUG] ReferenceTransformer: "output.bar (expand)" references: [var.foo var.foo]
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.ReferenceTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.AttachDependenciesTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.AttachDependenciesTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.attachDataResourceDependsOnTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.attachDataResourceDependsOnTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.DestroyEdgeTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.DestroyEdgeTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.pruneUnusedNodesTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.pruneUnusedNodesTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.TargetsTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.TargetsTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.ForcedCBDTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.ForcedCBDTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.CloseProviderTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.CloseProviderTransformer (no changes)
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.CloseRootModuleTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.CloseRootModuleTransformer with new graph:
  output.bar (expand) - *terraform.nodeExpandOutput
    var.foo - *terraform.NodeRootVariable
  root - *terraform.nodeCloseModule
    output.bar (expand) - *terraform.nodeExpandOutput
  var.foo - *terraform.NodeRootVariable
  ------
2023-09-09T23:10:36.779+0100 [TRACE] Executing graph transform *terraform.TransitiveReductionTransformer
2023-09-09T23:10:36.779+0100 [TRACE] Completed graph transform *terraform.TransitiveReductionTransformer (no changes)
2023-09-09T23:10:36.779+0100 [DEBUG] Starting graph walk: walkPlan
2023-09-09T23:10:36.779+0100 [TRACE] vertex "var.foo": starting visit (*terraform.NodeRootVariable)
2023-09-09T23:10:36.779+0100 [TRACE] NodeRootVariable: evaluating var.foo
2023-09-09T23:10:36.779+0100 [TRACE] prepareFinalInputVariableValue: preparing var.foo
2023-09-09T23:10:36.779+0100 [TRACE] prepareFinalInputVariableValue: var.foo has a default value
2023-09-09T23:10:36.779+0100 [TRACE] prepareFinalInputVariableValue: var.foo has no defined value
2023-09-09T23:10:36.779+0100 [TRACE] BuiltinEvalContext: Storing final value for variable var.foo
2023-09-09T23:10:36.779+0100 [TRACE] evalVariableValidations: no validation rules declared for var.foo, so skipping
2023-09-09T23:10:36.779+0100 [TRACE] vertex "var.foo": visit complete
2023-09-09T23:10:36.779+0100 [TRACE] vertex "output.bar (expand)": starting visit (*terraform.nodeExpandOutput)
2023-09-09T23:10:36.779+0100 [TRACE] vertex "output.bar (expand)": expanding dynamic subgraph
2023-09-09T23:10:36.779+0100 [TRACE] Expanding output: adding output.bar as *terraform.NodeApplyableOutput
2023-09-09T23:10:36.779+0100 [TRACE] vertex "output.bar (expand)": entering dynamic subgraph
2023-09-09T23:10:36.779+0100 [TRACE] vertex "output.bar": starting visit (*terraform.NodeApplyableOutput)
2023-09-09T23:10:36.780+0100 [TRACE] evalCheckRules: output.bar status is now StatusFail
2023-09-09T23:10:36.780+0100 [ERROR] vertex "output.bar" error: Module output value precondition failed
2023-09-09T23:10:36.780+0100 [TRACE] vertex "output.bar": visit complete, with errors
2023-09-09T23:10:36.780+0100 [TRACE] dag/walk: upstream of "root" errored, so skipping
2023-09-09T23:10:36.780+0100 [TRACE] vertex "output.bar (expand)": dynamic subgraph encountered errors: Module output value precondition failed
2023-09-09T23:10:36.780+0100 [ERROR] vertex "output.bar (expand)" error: Module output value precondition failed
2023-09-09T23:10:36.780+0100 [TRACE] vertex "output.bar (expand)": visit complete, with errors
2023-09-09T23:10:36.780+0100 [TRACE] dag/walk: upstream of "root" errored, so skipping
2023-09-09T23:10:36.780+0100 [DEBUG] TestFileRunner: completed plan for plan.tftest.hcl/plan
plan.tftest.hcl... [32mpass[0m[0m
  run "plan"... [32mpass[0m[0m
2023-09-09T23:10:36.780+0100 [TRACE] TestStateManager: cleaning up state for plan.tftest.hcl

[31mFailure![0m[0m 1 passed, 1 failed.
```

### Expected Behavior

This configuration has a single output `foo` that always fails its precondition. There are two test cases, one each for `command = plan` and `command = apply`, each of which expect the output to fail its precondition.

Both tests should pass.

### Actual Behavior

The "plan" test case passes, but "apply" fails. The precondition is observed to fail in the apply case.

```
$ terraform test
apply.tftest.hcl... fail
  run "apply"... fail
╷
│ Error: Module output value precondition failed
│
│   on main.tf line 10, in output "bar":
│   10:     condition     = false && var.foo
│     ├────────────────
│     │ var.foo is true
│
│ Always fails
╵
plan.tftest.hcl... pass
  run "plan"... pass

Failure! 1 passed, 1 failed.

```

### Steps to Reproduce

1. `terraform test`

### Additional Context

_No response_

### References

_No response_

--- a/internal/backend/local/test.go
+++ b/internal/backend/local/test.go
@@ -406,6 +406,7 @@ func (runner *TestFileRunner) run(run *moduletest.Run, file *moduletest.File, st
 
 	// Otherwise any error during the planning prevents our apply from
 	// continuing which is an error.
+	planDiags = run.ExplainExpectedFailures(planDiags)
 	run.Diagnostics = run.Diagnostics.Append(planDiags)
 	if planDiags.HasErrors() {
 		run.Status = moduletest.Error

--- a/internal/command/test_test.go
+++ b/internal/command/test_test.go
@@ -1145,3 +1145,119 @@ the current run block.
 		t.Errorf("should have deleted all resources on completion but left %v", provider.ResourceString())
 	}
 }
+
+func TestTest_ExpectedFailuresDuringPlanning(t *testing.T) {
+	td := t.TempDir()
+	testCopyDir(t, testFixturePath(path.Join("test", "expected_failures_during_planning")), td)
+	defer testChdir(t, td)()
+
+	provider := testing_command.NewProvider(nil)
+	view, done := testView(t)
+
+	c := &TestCommand{
+		Meta: Meta{
+			testingOverrides: metaOverridesForProvider(provider.Provider),
+			View:             view,
+		},
+	}
+
+	code := c.Run([]string{"-no-color"})
+	output := done(t)
+
+	if code == 0 {
+		t.Errorf("expected status code 0 but got %d", code)
+	}
+
+	expectedOut := `check.tftest.hcl... in progress
+  run "check_passes"... pass
+check.tftest.hcl... tearing down
+check.tftest.hcl... pass
+input.tftest.hcl... in progress
+  run "input_failure"... fail
+
+Warning: Expected failure while planning
+
+A custom condition within var.input failed during the planning stage and
+prevented the requested apply operation. While this was an expected failure,
+the apply operation could not be executed and so the overall test case will
+be marked as a failure and the original diagnostic included in the test
+report.
+
+input.tftest.hcl... tearing down
+input.tftest.hcl... fail
+output.tftest.hcl... in progress
+  run "output_failure"... fail
+
+Warning: Expected failure while planning
+
+  on output.tftest.hcl line 13, in run "output_failure":
+  13:     output.output,
+
+A custom condition within output.output failed during the planning stage and
+prevented the requested apply operation. While this was an expected failure,
+the apply operation could not be executed and so the overall test case will
+be marked as a failure and the original diagnostic included in the test
+report.
+
+output.tftest.hcl... tearing down
+output.tftest.hcl... fail
+resource.tftest.hcl... in progress
+  run "resource_failure"... fail
+
+Warning: Expected failure while planning
+
+A custom condition within test_resource.resource failed during the planning
+stage and prevented the requested apply operation. While this was an expected
+failure, the apply operation could not be executed and so the overall test
+case will be marked as a failure and the original diagnostic included in the
+test report.
+
+resource.tftest.hcl... tearing down
+resource.tftest.hcl... fail
+
+Failure! 1 passed, 3 failed.
+`
+	actualOut := output.Stdout()
+	if diff := cmp.Diff(actualOut, expectedOut); len(diff) > 0 {
+		t.Errorf("output didn't match expected:\nexpected:\n%s\nactual:\n%s\ndiff:\n%s", expectedOut, actualOut, diff)
+	}
+
+	expectedErr := `
+Error: Invalid value for variable
+
+  on main.tf line 2:
+   2: variable "input" {
+    ├────────────────
+    │ var.input is "bcd"
+
+input must contain the character 'a'
+
+This was checked by the validation rule at main.tf:5,3-13.
+
+Error: Module output value precondition failed
+
+  on main.tf line 33, in output "output":
+  33:     condition = strcontains(test_resource.resource.value, "d")
+    ├────────────────
+    │ test_resource.resource.value is "abc"
+
+input must contain the character 'd'
+
+Error: Resource postcondition failed
+
+  on main.tf line 16, in resource "test_resource" "resource":
+  16:       condition = strcontains(self.value, "b")
+    ├────────────────
+    │ self.value is "acd"
+
+input must contain the character 'b'
+`
+	actualErr := output.Stderr()
+	if diff := cmp.Diff(actualErr, expectedErr); len(diff) > 0 {
+		t.Errorf("output didn't match expected:\nexpected:\n%s\nactual:\n%s\ndiff:\n%s", expectedErr, actualErr, diff)
+	}
+
+	if provider.ResourceCount() > 0 {
+		t.Errorf("should have deleted all resources on completion but left %v", provider.ResourceString())
+	}
+}

--- a/internal/command/testdata/test/expected_failures_during_planning/check.tftest.hcl
+++ b/internal/command/testdata/test/expected_failures_during_planning/check.tftest.hcl
@@ -1,0 +1,16 @@
+
+run "check_passes" {
+
+  variables {
+    input = "abd"
+  }
+
+  # Checks are a little different, as they only produce warnings. So in this
+  # case we actually expect the whole run block to be fine. It'll produce
+  # warnings during the plan, still execute the apply operation, and then
+  # validate the check block failed during the apply stage.
+  expect_failures = [
+    check.cchar,
+  ]
+
+}

--- a/internal/command/testdata/test/expected_failures_during_planning/input.tftest.hcl
+++ b/internal/command/testdata/test/expected_failures_during_planning/input.tftest.hcl
@@ -1,0 +1,16 @@
+
+run "input_failure" {
+
+  variables {
+    input = "bcd"
+  }
+
+  # While we do expect var.input to fail, we are asking this run block to
+  # execute an apply operation. It can't do that because our custom condition
+  # fails during the planning stage as well. Our test is going to make sure we
+  # add the helpful warning diagnostic explaining this.
+  expect_failures = [
+    var.input,
+  ]
+
+}

--- a/internal/command/testdata/test/expected_failures_during_planning/main.tf
+++ b/internal/command/testdata/test/expected_failures_during_planning/main.tf
@@ -1,0 +1,37 @@
+
+variable "input" {
+  type = string
+
+  validation {
+    condition = strcontains(var.input, "a")
+    error_message = "input must contain the character 'a'"
+  }
+}
+
+resource "test_resource" "resource" {
+  value = var.input
+
+  lifecycle {
+    postcondition {
+      condition = strcontains(self.value, "b")
+      error_message = "input must contain the character 'b'"
+    }
+  }
+}
+
+check "cchar" {
+  assert {
+    condition = strcontains(test_resource.resource.value, "c")
+    error_message = "input must contain the character 'c'"
+  }
+}
+
+output "output" {
+  value = test_resource.resource.value
+
+  precondition {
+    condition = strcontains(test_resource.resource.value, "d")
+    error_message = "input must contain the character 'd'"
+  }
+}
+

--- a/internal/command/testdata/test/expected_failures_during_planning/output.tftest.hcl
+++ b/internal/command/testdata/test/expected_failures_during_planning/output.tftest.hcl
@@ -1,0 +1,16 @@
+
+run "output_failure" {
+
+  variables {
+    input = "abc"
+  }
+
+  # While we do expect output.output to fail, we are asking this run block to
+  # execute an apply operation. It can't do that because our custom condition
+  # fails during the planning stage as well. Our test is going to make sure we
+  # add the helpful warning diagnostic explaining this.
+  expect_failures = [
+    output.output,
+  ]
+
+}

--- a/internal/command/testdata/test/expected_failures_during_planning/resource.tftest.hcl
+++ b/internal/command/testdata/test/expected_failures_during_planning/resource.tftest.hcl
@@ -1,0 +1,16 @@
+
+run "resource_failure" {
+
+  variables {
+    input = "acd"
+  }
+
+  # While we do expect test_resource.resource to fail, we are asking this run
+  # block to execute an apply operation. It can't do that because our custom
+  # condition fails during the planning stage as well. Our test is going to make
+  # sure we add the helpful warning diagnostic explaining this.
+  expect_failures = [
+    test_resource.resource,
+  ]
+
+}

--- a/internal/moduletest/run.go
+++ b/internal/moduletest/run.go
@@ -108,6 +108,122 @@ func (run *Run) GetReferences() ([]*addrs.Reference, tfdiags.Diagnostics) {
 	return references, diagnostics
 }
 
+// ExplainExpectedFailures is similar to ValidateExpectedFailures except it
+// looks for any diagnostics produced by custom conditions and are included in
+// the expected failures and adds an additional explanation that clarifies the
+// expected failures are being ignored this time round.
+//
+// Generally, this function is used during an `apply` operation to explain that
+// an expected failure during the planning stage will still result in the
+// overall test failing as the plan failed and we couldn't even execute the
+// apply stage.
+func (run *Run) ExplainExpectedFailures(originals tfdiags.Diagnostics) tfdiags.Diagnostics {
+
+	// We're going to capture all the checkable objects that are referenced
+	// from the expected failures.
+	expectedFailures := addrs.MakeMap[addrs.Referenceable, bool]()
+	sourceRanges := addrs.MakeMap[addrs.Referenceable, tfdiags.SourceRange]()
+
+	for _, traversal := range run.Config.ExpectFailures {
+		// Ignore the diagnostics returned from the reference parsing, these
+		// references will have been checked earlier in the process by the
+		// validate stage so we don't need to do that again here.
+		reference, _ := addrs.ParseRefFromTestingScope(traversal)
+		expectedFailures.Put(reference.Subject, false)
+		sourceRanges.Put(reference.Subject, reference.SourceRange)
+	}
+
+	var diags tfdiags.Diagnostics
+	for _, diag := range originals {
+		if diag.Severity() == tfdiags.Warning {
+			// Then it's fine, the test will carry on without us doing anything.
+			diags = diags.Append(diag)
+			continue
+		}
+
+		if rule, ok := addrs.DiagnosticOriginatesFromCheckRule(diag); ok {
+
+			var rng *hcl.Range
+			expected := false
+			switch rule.Container.CheckableKind() {
+			case addrs.CheckableOutputValue:
+				addr := rule.Container.(addrs.AbsOutputValue)
+				if !addr.Module.IsRoot() {
+					// failures can only be expected against checkable objects
+					// in the root module. This diagnostic will be added into
+					// returned set below.
+					break
+				}
+				if expectedFailures.Has(addr.OutputValue) {
+					expected = true
+					rng = sourceRanges.Get(addr.OutputValue).ToHCL().Ptr()
+				}
+
+			case addrs.CheckableInputVariable:
+				addr := rule.Container.(addrs.AbsInputVariableInstance)
+				if !addr.Module.IsRoot() {
+					// failures can only be expected against checkable objects
+					// in the root module. This diagnostic will be added into
+					// returned set below.
+					break
+				}
+				if expectedFailures.Has(addr.Variable) {
+					expected = true
+				}
+
+			case addrs.CheckableResource:
+				addr := rule.Container.(addrs.AbsResourceInstance)
+				if !addr.Module.IsRoot() {
+					// failures can only be expected against checkable objects
+					// in the root module. This diagnostic will be added into
+					// returned set below.
+					break
+				}
+				if expectedFailures.Has(addr.Resource) {
+					expected = true
+				}
+
+				if expectedFailures.Has(addr.Resource.Resource) {
+					expected = true
+				}
+
+			case addrs.CheckableCheck:
+				// Check blocks only produce warnings so this branch shouldn't
+				// ever be triggered anyway.
+			default:
+				panic("unrecognized CheckableKind: " + rule.Container.CheckableKind().String())
+			}
+
+			if expected {
+				// Then this diagnostic was produced by a custom condition that
+				// was expected to fail. But, it happened at the wrong time (eg.
+				// we're trying to run an apply operation and this condition
+				// failed during the plan so the overall test operation still
+				// fails).
+				//
+				// We'll add a warning diagnostic explaining why the overall
+				// test is still failing even though the error was expected, and
+				// then add the original error into our diagnostics directly
+				// after.
+
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagWarning,
+					Summary:  "Expected failure while planning",
+					Detail:   fmt.Sprintf("A custom condition within %s failed during the planning stage and prevented the requested apply operation. While this was an expected failure, the apply operation could not be executed and so the overall test case will be marked as a failure and the original diagnostic included in the test report.", rule.Container.String()),
+					Subject:  rng,
+				})
+				diags = diags.Append(diag)
+				continue
+			}
+		}
+
+		// Otherwise, there is nothing special about this diagnostic so just
+		// carry it through.
+		diags = diags.Append(diag)
+	}
+	return diags
+}
+
 // ValidateExpectedFailures steps through the provided diagnostics (which should
 // be the result of a plan or an apply operation), and does 3 things:
 //  1. Removes diagnostics that match the expected failures from the config.

